HW2/P5/physics.pyx

```diff
@@ -69,17 +81,45 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+    #find the grid coordinates for the center of the ball
+    # considered 
+    X_cent = <int>(XY[i, 0] / grid_spacing)
+    Y_cent = <int>(XY[i, 1] / grid_spacing)
+    # find the max and min grid coordinates to track
+    # we check the objects 2 grid squares away
+    # therefore 3 grid points away
+    X_max =  min(X_cent+3, Grid.shape[0] - 1)
+    X_min =  max(X_cent-3, 0)
+    # find the max and min grid coordinates to track
+    # we check the objects 2 grid squares away
+    # therefore 3 grid points away
+    Y_max =  min(Y_cent+3, Grid.shape[1] - 1)
+    Y_min =  max(Y_cent-3, 0)
+
+    # loop over the neighbors to check
+    # if there is any ball
+    for X_check in range(X_min, X_max+1):
+        for Y_check in range(Y_min, Y_max):
+            j = Grid[X_check, Y_check]
+            # check only one side (avoid double counting)
+            # don't check when there is no ball (j = int_max not -1 like said in the driver)
+            if j>i and j != int_max :
+                XY2 = &(XY[j, 0])
+                V2 = &(V[j, 0])
+                if overlapping(XY1, XY2, R):
+                    # SUBPROBLEM 4: Add locking
+                    # to avoir deadlock, we access them in the same order
+                    # here it is smaller first
+                    if not moving_apart(XY1, V1, XY2, V2):
+                        acquire(&(locks[i]))
```

Comment:
You should acquire the first lock before you get the values of XY1 & V1 and the second lock before you get the values of XY2 & V2, otherwise they might have changed between when you read the values and when you do the collision. I think this is why you're not seeing much of a slowdown with locking while most everyone else is - you're not locking for long enough to actually ensure accuracy.